core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java

```diff
@@ -99,57 +117,104 @@ private boolean isEndOfStream() {
     return (!activeBuffer.hasRemaining() && !readAheadBuffer.hasRemaining() && endOfStream);
   }
 
-  private void readAsync(final ByteBuffer byteBuffer) throws IOException {
+  private void checkReadException() throws IOException {
+    if (readAborted) {
+      Throwables.propagateIfPossible(readException, IOException.class);
+      throw new IOException(readException);
+    }
+  }
+
+  /** Read data from underlyingInputStream to readAheadBuffer asynchronously. */
+  private void readAsync() throws IOException {
     stateChangeLock.lock();
-    final byte[] arr = byteBuffer.array();
+    final byte[] arr = readAheadBuffer.array();
     try {
       if (endOfStream || readInProgress) {
         return;
       }
-      byteBuffer.position(0);
-      byteBuffer.flip();
+      checkReadException();
+      readAheadBuffer.position(0);
+      readAheadBuffer.flip();
       readInProgress = true;
     } finally {
       stateChangeLock.unlock();
     }
     executorService.execute(new Runnable() {
+
       @Override
       public void run() {
+        stateChangeLock.lock();
+        try {
+          if (isClosed) {
+            readInProgress = false;
+            return;
+          }
+          // Flip this so that the close method will not close the underlying input stream when we
+          // are reading.
+          isReading = true;
```

Comment:
Why do we need this variable? Can we use `readInProgress` instead?

Reply:
`readInProgress == true` means the task will be submitted. But it may not run if the executor service is shut down. I changed `shutdown` to `shutdownNow` in order to interrupt the blocking read.

Reply:
I see. Instead of doing that should we just call `waitForAsyncReadComplete` in `close()` function call so that we can safely close the input stream? It would clean up the code a lot and make it less complicated.

Reply:
`waitForAsyncReadComplete` can be interrupted as well.